### PR TITLE
Clarify meaning of `training` parameter in XGBoosterPredict()

### LINKED ISSUE #5601: v1.1.0rc1 breaking API change: when should `training` be true in XGBoosterPredict(handle, dmat, option_mask, ntree_limit, training, ...)
I'm trying to update downstream libraries for v1.1.0rc1. There was a breaking change in the c_api.h to have an additional `training bool` parameter in this code change:

https://github.com/hcho3/xgboost/commit/f100b8d87871adaa693ea84548d69a598672d6f9
~~~
* Add `training` parameter for prediction method.
* [Breaking]: Add `training` to C API.
~~~

I'm looking at the patch but cannot tell when training should be true or false in the XGBoosterPredict() call. Could somebody offer some guidance?

I read that 
~~~
* \param training Whether the prediction value is used for training.
~~~

but that doesn't really tell me enough to understand what the consequences of training true (1) will be.

Thanks for you help.

### PR DESCRIPTION
**Pull request by glycerine**
fixes #5601 . Pre-empts questions about the breaking API change for v1.1.0 that downstream maintainers will have.

### CODE REVIEW

**Comment by hcho3** [comment]
@glycerine I simplified the explanation in the docstring. Can you review and see if it makes sense to you?

**Comment by glycerine** [comment]
@hcho3 LGTM. Thanks for adding the clarification about custom objectives.

**Comment by trivialfis** [comment]
This way the parameter is general enough to be reused in the future instead of being a dart only parameter.

**Comment by hcho3** [comment]
@trivialfis Can you suggest a good way to document the meaning of `training` for people who use the C API function? Right now, it is not clear when `training` parameter should be set to 1.

**Comment by hcho3** [comment]
```
 * \param training Whether the prediction function is used as part of a training loop.
 *    If you are running prediction on a model that's finished training, set `training` to false.
```

How about this?

**Comment by glycerine** [comment]
YAGNI applies. I would say don't document it as general-purpose until it actual is. It is already used here in R in the general sense, even when DART may not be used.

https://github.com/dmlc/xgboost/blob/master/R-package/R/utils.R#L148

Is there harm is setting to true when not under DART?

**Comment by trivialfis** [comment]
Maybe we can give `dart` as an example.

**Comment by glycerine** [comment]
> If you are running prediction on a model that's finished training, set `training` to false.

A little too sparse. I would want a little more guidance about when it is okay to set it to true.

**Comment by trivialfis** [comment]
@glycerine How about this, if the prediction result is used for calculating gradient for training, set it to true, otherwise set it to false.

**Comment by trivialfis** [comment]
Also, it's a breaking change in 1.0.0.

**Comment by hcho3** [comment]
```
* \param training Whether the prediction function is used as part of a training loop.
*    1. If you are running prediction on a model that's finished training, set training=false.
*    2. When should you set training=true?
*       Normally, you don't need to use XGBoosterPredict() in your training loop, since you are
*       probably using XGBoosterUpdateOneIter(), which calls XGBoosterPredict() internally. The one
*       case you should use XGBoosterPredict() in your training loop is when you wrote a customized
*       objective. In this case, call XGBoosterPredict() with training=true and then invoke your
*       customized objective fobj(predicted_label, true_label) to obtain the gradient pairs.
```

I tried elaborating the design behind the `training` parameter.

**Comment by glycerine** [comment]
~~~
* \param training Whether the prediction function is used as part of a training loop.
*    1. If you are running prediction on a model that's finished training, set training=false.
*    2. When should you set training=true?
*       Normally, you don't need to use XGBoosterPredict() in your training loop, since you are
*       probably using XGBoosterUpdateOneIter(), which calls XGBoosterPredict() internally. The one
*       case you should use XGBoosterPredict() in your training loop is when you wrote a customized
*       objective. In this case, call XGBoosterPredict() with training=true and then invoke your
*       customized objective fobj(predicted_label, true_label) to obtain the gradient pairs.
~~~

@hcho3 I like it. Very clear and helpful.

**Comment by trivialfis** [comment]
> *    1. If you are running prediction on a model that's finished training, set training=false.

I would omit this item.  We do prediction during training to use external metrics.

**Comment by glycerine** [comment]
> I would omit this item. We do prediction during training to use external metrics.

@trivialfis That is the most important piece of guidance. Without that, we are lost. 

You really must describe detail when to set it true, and when to set it false. Give examples. Be precise. Without the line above, it is far too vague, and you will get tons of questions.

**Comment by hcho3** [comment]
```
* \param training Whether the prediction function is used as part of a training loop.
*    When should you set training=true? Set training=true if 1) you are writing your training loop 
*    and 2) you are using a customized objective function to obtain gradient pairs.
*    Normally, you don't need to use XGBoosterPredict() in your training loop, since you are
*    probably using XGBoosterUpdateOneIter(), which calls XGBoosterPredict() internally. The one
*    case you should use XGBoosterPredict() in your training loop is when you wrote a customized
*    objective. In this case, call XGBoosterPredict() with training=true and then invoke your
*    customized objective fobj(predicted_label, true_label) to obtain the gradient pairs.
```

**Comment by glycerine** [comment]
great. This is good in that it give guidance about when to use training=true.

It would also be good to add what change happens when you set training true, and that shows the intent. I'm guessing it has to do with dropout or not, as one example.

**Comment by trivialfis** [comment]
Prediction can be run in 2 scenarios:
- Normal model inference, given a X, obtain the y_p from model.
- Obtain the prediction for computing gradients.  For example `dart` booster performs dropout during training, and the prediction result will be different from the one obtained by normal inference step.

The `training` parameter is used to distinguish these 2 scenarios.  Set to `false` for the first scenario and `true` for the second one.

**Comment by glycerine** [comment]
@trivialfis Excellent.

File: include/xgboost/c_api.h [hunk 1]
```diff
@@ -418,7 +418,14 @@ XGB_DLL int XGBoosterEvalOneIter(BoosterHandle handle,
...
  *    when the parameter is set to 0, we will use all the trees
- * \param training Whether the prediction value is used for training.
+ * \param training Whether the prediction function is used as part of a training loop.
+ *    Prediction can be run in 2 scenarios:
+ *    1. Given data matrix X, obtain prediction y_pred from the model.
+ *    2. Obtain the prediction for computing gradients. For example, DART booster performs dropout
+ *       during training, and the prediction result will be different from the one obtained by normal
+ *       inference step due to dropped trees.
+ *    Set training=false for the first scenario. Set training=true for the second scenario.
+ *    The second scenario applies when you are defining a custom objective function.
```
Comment (hcho3) [inline include/xgboost/c_api.h:428]: Two scenarios are now listed. DART is used as an example. Also mentioning custom objective as an example of Scenario 2.

Comment (glycerine) [reply include/xgboost/c_api.h:428]: looks great to me.
